From c1d84c205448806cbd819831acc54ae633d8379f Mon Sep 17 00:00:00 2001
From: ming_qian <ming.qian@nxp.com>
Date: Wed, 3 Jul 2019 09:50:26 +0800
Subject: [PATCH 07/52] MLK-22141: VPU Decoder: fix consumed length calculation
 error

when check whether the rptr is cross over,
the stream size must be taken into account.

Signed-off-by: ming_qian <ming.qian@nxp.com>
Reviewed-by: Shijie Qin <shijie.qin@nxp.com>
Reviewed-by: Zhou Peng <eagle.zhou@nxp.com>
---
 drivers/mxc/vpu_malone/vpu_b0.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/mxc/vpu_malone/vpu_b0.c b/drivers/mxc/vpu_malone/vpu_b0.c
index 867818106eb9..a2f4dc3ba7fc 100644
--- a/drivers/mxc/vpu_malone/vpu_b0.c
+++ b/drivers/mxc/vpu_malone/vpu_b0.c
@@ -3460,10 +3460,19 @@ static void add_buffer_to_queue(struct queue_data *q_data, struct vb2_data_req *
 static u32 correct_consumed_length(struct vpu_ctx *ctx,
 				u32 consumed_pic_bytesused)
 {
+	long total_read_bytes;
 	long delta;
 	u32 circle_count;
+	u32 stream_size;
+	pSTREAM_BUFFER_DESCRIPTOR_TYPE pStrBufDesc;
 
-	delta = ctx->total_write_bytes - ctx->total_consumed_bytes;
+	pStrBufDesc = get_str_buffer_desc(ctx);
+	stream_size = got_used_space(pStrBufDesc->wptr,
+					pStrBufDesc->rptr,
+					pStrBufDesc->start,
+					pStrBufDesc->end);
+	total_read_bytes = ctx->total_write_bytes - stream_size;
+	delta = total_read_bytes - ctx->total_consumed_bytes;
 	if (delta < ctx->stream_buffer.dma_size)
 		return consumed_pic_bytesused;
 
-- 
2.17.1

